iso.bbx

```diff
@@ -208,14 +214,26 @@
 % \renewcommand*{\mkbibacro}[1]{#1}
 % Applies to DOI, ISBN, ISSN, ISAN, ISMN, ISRN, ISWC
 
-% We must handle more than one isbn and issn
+% If we really must handle more than one ISBN and ISSN.
+%
+% Typically there could be different ISBN / ISSN for printed and electronic
+% media versions, but the standard says that the specific copy or instance of
+% the document should be cited, so the author should choose the appropriate
+% number in the bib file. Otherwise include comma-separated list of numbers in
+% the bib field 'isbn' or 'issn'.
+
+% Separator value between more ISBNs / ISSNs is handled by '\stdidentifierspunct'.
+% Default format is the name of identifier + space + number:
+% e.g. 'ISBN 978-3-642-21665-7'
 \DeclareFieldFormat{isbn}{%
-  %\renewcommand*{\do}[1]{\mkbibacro{ISBN}\addspace#1\adddot\space}%
-  \mkbibacro{ISBN}\addspace#1}%
+  \def\do##1{\stdidentifierspunct\mkbibacro{ISBN}\space##1}%
```

Comment:
With that definition `` also applies before the first ISBN. Given the description says "between more standard identifiers" that was a bit unexpected to me and might not actually be desired.


```latex
\documentclass[english]{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{babel}
\usepackage{csquotes}

\usepackage[style=iso-authoryear, thesisinfoinnotes=false, backend=biber]{biblatex}

\renewcommand\stdidentifierspunct{XXX}

\addbibresource{biblatex-examples.bib}

\begin{document}
\autocite{cms}
\printbibliography
\end{document}
```

Reply:
True, I will dig in the details of this commit a bit:
* the original `\do` command was not working at all with `\renewcommand` syntax (I supposed it should)
* I used `\def` clause, which I remember you are not fond of it, but only this works for me in this case
* I relied on the `\stdidentifierspunct` will not be printed if it contains punctuation symbol(s), because it follows some other punctuation from the previous block, but it's true that someone might use sth like `or the other ISBN` as a "punctuation"
* if `\stdidentifierspunct` is at the end of `\do` command, it also applies if there is only one ISBN, which is also not desired
* I couldn't figure out how to easily check if there is more than one number in the list
* now I realized we can change type of `isbn` from the `field` to `list` entry type and then check with `\ifmoreitems` for multiple ISBNs, can't we?

I would really like to do it the right way.

Reply:
* It shouldn't really matter if we use `\def` or `\renewcommand`
    ```latex
    \renewcommand*{\do}[1]{\stdidentifierspunct\mkbibacro{ISBN}\space##1}%
   ```
    works just as well.
* While I usually prefer not to use `\def`, I think it is OK with `\do`. The `etoolbox` documentation uses `\renewcommand*{\do}`, but all instances but one in the `biblatex` core say `\def\do`
* Yeah, you'd usually want to know the number of items in advance and it would be nice to have a running counter as well. I don't know how to do that without looping over the field first just for counting and looping over it for a second time for the actual typesetting. In this case it is enough to cheat a bit
    ```latex
    \documentclass[british]{article}
    \usepackage[T1]{fontenc}
    \usepackage[utf8]{inputenc}
    \usepackage{babel}
    \usepackage{csquotes}

    \usepackage[style=iso-authoryear, backend=biber]{biblatex}

    \renewcommand\stdidentifierspunct{XXX}

    \DeclareFieldFormat{isbn}{%
      \def\do##1{\mkbibacro{ISBN}\space##1%
        \def\do####1{\stdidentifierspunct\mkbibacro{ISBN}\space####1}}%
      \docsvfield{isbn}%
    }

    \begin{filecontents}[force]{\jobname.bib}
    @book{appleby,
      author  = {Humphrey Appleby},
      title   = {On the Importance of the Civil Service},
      date    = {1980},
      isbn    = {1-234,456-7},
    }
    \end{filecontents}
    \addbibresource{\jobname.bib}


    \begin{document}
    \cite{appleby}
    \printbibliography
    \end{document}
    ```
* Yes, a list would be the best solution if you want multiple ISBNs.
    ```latex
    \documentclass[british]{article}
    \usepackage[T1]{fontenc}
    \usepackage[utf8]{inputenc}
    \usepackage{babel}
    \usepackage{csquotes}

    \begin{filecontents}[force]{iso-authoryear.dbx}
    \ProvidesFile{iso-authoryear.dbx}[2017/04/25 v0.3.2 biblatex data model extension]
    \RequireBiber[3]
    \DeclareDatamodelFields[type=list,datatype=name]{supervisor}
    \DeclareDatamodelEntryfields[thesis]{supervisor}


    \DeclareDatamodelFields[type=list, datatype=literal]{isbn}
    \end{filecontents}

    \usepackage[style=iso-authoryear, backend=biber]{biblatex}

    \newbibmacro*{list:stdidentifierspunct}[1]{%
      \ifnumgreater{\value{listcount}}{\value{liststart}}
        {\stdidentifierspunct}
        {}}

    \DeclareListFormat{isbn}{%
      \usebibmacro{list:stdidentifierspunct}%
      \mkbibacro{ISBN}\space#1%
      \usebibmacro{list:andothers}}


    \newbibmacro*{identifier}{%
      \iftoggle{bbx:isbn}
        {\printfield{isan}%
         \newunit
         \printlist{isbn}%
         \newunit
         \printfield{ismn}%
         \newunit
         \printfield{isrn}%
         \newunit
         \printfield{issn}%
         \newunit
         \printfield{iswc}%
         \newunit}
        {}%
    }

    \begin{filecontents}[force]{\jobname.bib}
    @book{appleby,
      author  = {Humphrey Appleby},
      title   = {On the Importance of the Civil Service},
      date    = {1980},
      isbn    = {1-234 and 456-7},
    }
    \end{filecontents}
    \addbibresource{\jobname.bib}


    \begin{document}
    \cite{appleby}
    \printbibliography
    \end{document}
    ```
    Would work, but uses the usual list format with `and` and not with commas, which might be a bit of a backwards compatibility issue.

Reply:
* It shouldn't really matter if we use `\def` or `\renewcommand`
    ```latex
    \renewcommand*{\do}[1]{\stdidentifierspunct\mkbibacro{ISBN}\space##1}%
   ```
    works just as well.
* While I usually prefer not to use `\def`, I think it is OK with `\do`. The `etoolbox` documentation uses `\renewcommand*{\do}`, but all instances but one in the `biblatex` core say `\def\do`
* Yeah, you'd usually want to know the number of items in advance and it would be nice to have a running counter as well. I don't know how to do that without looping over the field first just for counting and looping over it for a second time for the actual typesetting. In this case it is enough to cheat a bit
    ```latex
    \documentclass[british]{article}
    \usepackage[T1]{fontenc}
    \usepackage[utf8]{inputenc}
    \usepackage{babel}
    \usepackage{csquotes}

    \usepackage[style=iso-authoryear, backend=biber]{biblatex}

    \renewcommand\stdidentifierspunct{XXX}

    \DeclareFieldFormat{isbn}{%
      \def\do##1{\mkbibacro{ISBN}\space##1%
        \def\do####1{\stdidentifierspunct\mkbibacro{ISBN}\space####1}}%
      \docsvfield{isbn}%
    }

    \begin{filecontents}[force]{\jobname.bib}
    @book{appleby,
      author  = {Humphrey Appleby},
      title   = {On the Importance of the Civil Service},
      date    = {1980},
      isbn    = {1-234,456-7},
    }
    \end{filecontents}
    \addbibresource{\jobname.bib}


    \begin{document}
    \cite{appleby}
    \printbibliography
    \end{document}
    ```
* Yes, a list would be the best solution if you want multiple ISBNs.
    ```latex
    \documentclass[british]{article}
    \usepackage[T1]{fontenc}
    \usepackage[utf8]{inputenc}
    \usepackage{babel}
    \usepackage{csquotes}

    \begin{filecontents}[force]{iso-authoryear.dbx}
    \ProvidesFile{iso-authoryear.dbx}[2017/04/25 v0.3.2 biblatex data model extension]
    \RequireBiber[3]
    \DeclareDatamodelFields[type=list,datatype=name]{supervisor}
    \DeclareDatamodelEntryfields[thesis]{supervisor}


    \DeclareDatamodelFields[type=list, datatype=literal]{isbn}
    \end{filecontents}

    \usepackage[style=iso-authoryear, backend=biber]{biblatex}

    \newbibmacro*{list:stdidentifierspunct}[1]{%
      \ifnumgreater{\value{listcount}}{\value{liststart}}
        {\stdidentifierspunct}
        {}}

    \DeclareListFormat{isbn}{%
      \usebibmacro{list:stdidentifierspunct}%
      \mkbibacro{ISBN}\space#1%
      \usebibmacro{list:andothers}}


    \newbibmacro*{identifier}{%
      \iftoggle{bbx:isbn}
        {\printfield{isan}%
         \newunit
         \printlist{isbn}%
         \newunit
         \printfield{ismn}%
         \newunit
         \printfield{isrn}%
         \newunit
         \printfield{issn}%
         \newunit
         \printfield{iswc}%
         \newunit}
        {}%
    }

    \begin{filecontents}[force]{\jobname.bib}
    @book{appleby,
      author  = {Humphrey Appleby},
      title   = {On the Importance of the Civil Service},
      date    = {1980},
      isbn    = {1-234 and 456-7},
    }
    \end{filecontents}
    \addbibresource{\jobname.bib}


    \begin{document}
    \cite{appleby}
    \printbibliography
    \end{document}
    ```
    Would work, but uses the usual list format with `and` and not with commas, which might be a bit of a backwards compatibility issue.

Reply:
I pushed that cheat version of `\DeclareFieldFormat{isbn}` into master, since as the standard says, there shouldn't be more ISBNs for the one specific entry, so use this only as the last fallback and do not temp users to adjust bib entries to have more ISBNs.